zookeeper-server/src/main/java/org/apache/zookeeper/backup/NFSBackupStorage.java

```diff
@@ -0,0 +1,324 @@
+              + " to backup storage. The message of the IOException is: " + e.getMessage());
+    } finally {
+      inputStream.close();
+      if (outputStream != null) {
+        outputStream.close();
+      }
+    }
+
+    // Rename the destination temporary file to destination file name provided by caller
+    try {
+      Files.move(Paths.get(backupTempFilePath), Paths.get(backupFilePath),
+          StandardCopyOption.REPLACE_EXISTING);
+    } catch (IOException e) {
+      throw new IOException(
+          "IOException is caught during copying temp file to destination file " + backupFile
+              .getPath() + ". The message of the IOException is: " + e.getMessage());
+    }
+  }
+
+  @Override
+  public void copyToLocalStorage(File srcName, File destFile) throws IOException {
+    InputStream inputStream;
+    OutputStream outputStream = null;
+
+    // Create input stream from the source file in backup storage
+    String backupFilePath = constructBackupFilePath(srcName.getName());
+    File backupFile = new File(backupFilePath);
+    try {
+      inputStream = new FileInputStream(backupFile);
+    } catch (FileNotFoundException e) {
+      if (!backupFile.exists() || backupFile.isDirectory()) {
+        throw new BackupException("The backed up file with the file path " + backupFilePath
+            + " does exists or is a directory, could not copy to local storage.");
+      } else {
+        throw new IOException("Cannot open source file in backup storage" + backupFilePath
+            + " , cannot be copied to local storage.");
+      }
+    }
+
+    // Prepare in local storage: create parent dirs if not exist, create the destination file
+    destFile.getParentFile().mkdirs();
+    if (!destFile.createNewFile()) {
+      if (destFile.exists()) {
+        throw new BackupException("A local storage file with the file path " + destFile.getPath()
+            + " already exists, could not create a new one.");
+      } else {
+        throw new IOException("Failed to create a file at destination path: " + destFile.getPath()
+            + ", stopping the streaming.");
+      }
+    }
+
+    // Create output stream on the destination file and start streaming data
+    try {
+      outputStream = new FileOutputStream(destFile);
+
+      byte[] buffer = new byte[1024];
+      int lengthRead;
+      while ((lengthRead = inputStream.read(buffer)) > 0) {
+        outputStream.write(buffer, 0, lengthRead);
+        outputStream.flush();
+      }
+    } catch (FileNotFoundException e) {
+      throw new IOException(
+          "The destination local file " + destFile.getPath() + " could not be opened for write.");
+    } catch (IOException e) {
+      throw new IOException(
+          "IOException is caught during data streaming while copying file " + backupFilePath
+              + " to local storage. The message of the IOException is: " + e.getMessage());
```

Comment:
We want to log these exceptions. For example, the original exception message for (FileNotFoundException e) is lost because you just re-throw an IOException.

Also, should we use BackupException (instead of just using IOException) where appropriate? I reiterate, this is where defining exactly what type of exception should be thrown would be helpful.

Reply:
Has this been resolved?